lib/Serializer.ts

```diff
@@ -0,0 +1,120 @@
+		}
+
+		return serialized;
+	}, serialized);
+};
+
+const excludeHandler = (document: ModelType<Document>, excludeRules: string[], serialized: ObjectType): ObjectType => {
+	return excludeRules.reduce((serialized: ObjectType, key: string) => {
+		if (Object.keys(serialized).includes(key)) {
+			delete serialized[key];
+		}
+
+		return serialized;
+	}, serialized);
+};
+
+const validateName = (name: string): void => {
+	if (!name || typeof name !== "string") {
+		throw new Error("Field name is required and should be of type string");
+	}
+};
+
+const validateOptions = (options): void => {
+	if (!options || !(Array.isArray(options) || typeof options === "object")) {
+		throw new Error("Field options is required and should be an object or array");
+	}
+};
+
+const cleanAndValidateDocumentsArray = (documentsArray: ModelType<Document>[]): ModelType<Document>[] => {
+	if (!documentsArray || !Array.isArray(documentsArray)) {
+		throw new Error("documentsArray must be an array of document objects");
+	}
+
+	return documentsArray.filter((doc) => typeof doc.serialize === "function");
+};
+
+export class Serializer {
+	#serializers: any;
+	#defaultSerializer: string;
+
+	constructor() {
+		this.#serializers = {_default: defaultSerializer};
+		this.#defaultSerializer = "_default";
+	}
+
+	add(name: string, options): void {
+		validateName(name);
+		validateOptions(options);
+		this.#serializers[name] = options;
+	}
+
+	setDefault(name: string): void {
+		validateName(name);
+		if (Object.keys(this.#serializers).includes(name)) {
+			this.#defaultSerializer = name;
+		}
+	}
+
+	remove(name: string): void {
+		validateName(name);
+		if (Object.keys(this.#serializers).includes(name)) {
+			delete this.#serializers[name];
+		}
+		if (this.#defaultSerializer === name) {
+			this.#defaultSerializer = "_default";
+		}
+	}
+
+	_serializeMany(documentsArray: ModelType<Document>[] = [], nameOrOptions): ObjectType[] {
+		documentsArray = cleanAndValidateDocumentsArray(documentsArray);
+		return documentsArray.map((doc) => doc.serialize(nameOrOptions));
+	}
+
+	_serialize(document: ModelType<Document>, nameOrOptions = this.#defaultSerializer): ObjectType {
+		const inputType = typeof nameOrOptions;
+		let options;
+
+		if (inputType === "string") {
+			options = this.#serializers[nameOrOptions];
+		} else if (Array.isArray(nameOrOptions) || inputType === "object") {
+			options = nameOrOptions;
+		}
+
+		validateOptions(options);
+		if (Array.isArray(options)) {
+			return includeHandler(document, options, {});
+		}
+
+		let serialized: ObjectType = {};
+		if (options.include) {
+			serialized = includeHandler(document, options.include, serialized);
+		}
+		if (options.exclude) {
+			if (!options.include) {
+				serialized = {...document};
+			}
+			serialized = excludeHandler(document, options.exclude, serialized);
+		}
+		if (options.modify && typeof options.modify === "function") {
+			if (!options.include && !options.exclude) {
+				serialized = {...document};
```

Comment:
I think we should remove this and set it as the default value of this variable

Reply:
If im following your suggestion correctly, the reason the code is this way is if you didn't configure one of includes or excludes and only had a modify function, it would operate on the complete document  values. Otherwise continue to use the serialized values that have already ran.

Reply:
Ahh. I think I see what you mean. I still think there is a way to refactor this to have it work.